pallets/file-system/src/utils.rs

```diff
@@ -57,7 +61,84 @@ macro_rules! expect_or_err {
+        bsp_id: ProviderIdFor<T>,
+        file_key: MerkleHash<T>,
+    ) -> Result<BlockNumberFor<T>, QueryFileEarliestVolunteerBlockError>
+    where
+        T: frame_system::Config,
+        T::ThresholdType: From<u128>,
+    {
+        // Get the block number at which the storage request was created.
+        let (storage_request_block, fingerprint) = match <StorageRequests<T>>::get(&file_key) {
+            Some(storage_request) => (storage_request.requested_at, storage_request.fingerprint),
+            None => {
+                return Err(QueryFileEarliestVolunteerBlockError::StorageRequestNotFound);
+            }
+        };
+
+        let bsp_threshold = Self::compute_bsp_xor(
+            fingerprint
+                .as_ref()
+                .try_into()
+                .map_err(|_| QueryFileEarliestVolunteerBlockError::FailedToEncodeFingerprint)?,
+            &bsp_id
+                .encode()
+                .try_into()
+                .map_err(|_| QueryFileEarliestVolunteerBlockError::FailedToEncodeBsp)?,
+        )
+        .map_err(|_| QueryFileEarliestVolunteerBlockError::ThresholdArithmeticError)?;
+
+        // Compute the block number at which the BSP should send the volunteer request.
+        let volunteer_block_number =
+            match Self::compute_volunteer_block_number(bsp_threshold, storage_request_block) {
+                Ok(Some(volunteer_block_number)) => volunteer_block_number,
+                _ => Zero::zero(),
+            };
+
+        Ok(volunteer_block_number)
+    }
+
+    fn compute_volunteer_block_number(
+        bsp_threshold: T::ThresholdType,
+        storage_request_block: BlockNumberFor<T>,
+    ) -> Result<Option<BlockNumberFor<T>>, DispatchError>
+    where
+        T: frame_system::Config,
+        T::ThresholdType: From<u128>,
+    {
+        // Calculate the difference between BSP's threshold and the current threshold.
+        let threshold_diff = match bsp_threshold.checked_sub(&BspsAssignmentThreshold::<T>::get()) {
+            Some(diff) => diff,
+            None => {
+                // The BSP's threshold is less than the current threshold.
+                return Ok(None);
+            }
+        };
+
+        // Calculate the number of blocks required to exceed the difference in thresholds.
+        let blocks_to_wait =
+            match threshold_diff.checked_div(&T::AssignmentThresholdMultiplier::get()) {
+                Some(blocks) => blocks,
+                None => {
+                    return Err(Error::<T>::ThresholdArithmeticError.into());
+                }
+            };
+
+        // Compute the block number at which the BSP should send the volunteer request.
+        let blocks_to_wait: U256 = blocks_to_wait.into_u256();
+        let volunteer_block_number = storage_request_block.saturating_add(
+            BlockNumberFor::<T>::try_from(blocks_to_wait)
+                .map_err(|_| Error::<T>::FailedToConvertBlockNumber)?,
+        );
```

Comment:
This way of converting is fine and will work. If you're interested though, there is another way to do it without intermediately casting to a larget number type. You can find it [here](https://github.com/Moonsong-Labs/storage-hub/blob/main/pallets/proofs-dealer/src/utils.rs#L532). Basically it uses the inner type `StakeToBlockNumber` which has to implement the trait `Convert<BalanceFor<Self>, BlockNumberFor<Self>>`